Makefile.pre.in

```diff
@@ -739,6 +764,194 @@ regen-test-frozenmain: $(BUILDPYTHON)
 Programs/_testembed: Programs/_testembed.o $(LIBRARY_DEPS)
 	$(LINKCC) $(PY_CORE_LDFLAGS) $(LINKFORSHARED) -o $@ Programs/_testembed.o $(BLDLIBRARY) $(LIBS) $(MODLIBS) $(SYSLIBS)
 
+############################################################################
+# "Bootstrap Python" used to run deepfreeze.py
+
+BOOTSTRAP = ./_bootstrap_python
+BOOTSTRAP_HEADERS = \
+	Python/frozen_modules/importlib._bootstrap.h \
+	Python/frozen_modules/importlib._bootstrap_external.h \
+	Python/frozen_modules/zipimport.h
+
+Python/bootstrap_frozen.o: Python/bootstrap_frozen.c Include/cpython/import.h $(BOOTSTRAP_HEADERS)
+
+$(BOOTSTRAP): $(LIBRARY_OBJS_OMIT_FROZEN) \
+		Python/bootstrap_frozen.o Programs/python.o
+	$(LINKCC) $(PY_CORE_LDFLAGS) -o $@ $(LIBRARY_OBJS_OMIT_FROZEN) \
+		Python/bootstrap_frozen.o \
+		Programs/python.o \
+		$(LIBS) $(MODLIBS) $(SYSLIBS)
+
+############################################################################
+# Deepfreeze targets
+
+.PHONY: regen-deepfreeze
+regen-deepfreeze: $(DEEPFREEZE_OBJS)
+
+# BEGIN: deepfreeze modules
+
+Python/deepfreeze/importlib._bootstrap.c: Lib/importlib/_bootstrap.py Tools/scripts/deepfreeze.py $(BOOTSTRAP)
+	@echo "Deepfreezing Python/deepfreeze/importlib._bootstrap.c from Lib/importlib/_bootstrap.py"
+	@$(BOOTSTRAP) \
+		$(srcdir)/Tools/scripts/deepfreeze.py \
+		$(srcdir)/Lib/importlib/_bootstrap.py \
+		-m importlib._bootstrap \
+		-o $(srcdir)/Python/deepfreeze/importlib._bootstrap.c
```

Comment:
Makefiles are a complicated mess. There are some subtle issues in this code that breaks out-of-tree builds. The general rule is: any existing source file must be prefixed with ``$(srcdir)/``. Any generated file must not have the prefix, so they get placed in the build tree instead of the source tree.

You can also make use of make autovars here. The ``$@`` is the target name (e.g. ``Python/deepfreeze/importlib._bootstrap.c``) and ``$<`` is the left-most dependency (e.g. ``$(srcdir)/Lib/importlib/_bootstrap.py``).

```suggestion
Python/deepfreeze/importlib._bootstrap.c: $(srcdir)/Lib/importlib/_bootstrap.py $(srcdir)/Tools/scripts/deepfreeze.py $(BOOTSTRAP)
	@echo "Deepfreezing $@ from $<"
	@$(BOOTSTRAP) \
		$(srcdir)/Tools/scripts/deepfreeze.py \
		$< \
		-m importlib._bootstrap \
		-o $@
```

Can we teach ``deepfreeze.py`` to deduce the module name from the source name? It allow us to simplify the make rules even further. It might be possible to use a Makefile substitution rule here, too.

I'm dropping the code here for future reference:

```
DEEPFREEZE_PYS = \
	$(srcdir)/Lib/importlib/_bootstrap.py \
	$(srcdir)/Lib/importlib/_bootstrap_external.py \
	$(srcdir)/Lib/zipimport.py

# Lib/egg/spam.py -> egg/spam.c -> egg_spam.c -> Python/deepfreeze/egg_spam.c
DEEPFREEZE_CS = $(addprefix Python/deepfreeze/,$(subst /,.,$(patsubst $(srcdir)/Lib/%.py,%.c,$(DEEPFREEZE_PYS))))
```

Reply:
I found a better way to simplify your build rules, https://github.com/python/cpython/compare/main...tiran:deepfreeze-make

The updates rules use patterns (``%``), autovars (``$@`` and ``$<``), and target local variables (``target : KEY=VALUE``) to avoid repetition. I had to rename the target file name, though.

Reply:
I'm generating this part of `Makefile.pre.in` anyway and I'm not too keen on renaming the target files (or on using patterns), so I think I'll pass on that solution, but I think I've fixed the use of `$(srcdir)` and am now using `$@` and `$<` in the generated rules. I can now focus on fixing the 4 remaining tests.

But the real problem with this PR is that it slows down the build (much more than the current freeze approach) because whenever you touch a file (even files unrelated to freezing, say dictobject.c), the dependencies cause `_bootstrap_python` to be rebuilt and then to be used to regenerate all the deepfrozen .c files, which then have to be recompiled. And that takes several seconds and causes a lot of commands to scroll by, where previously it would just be recompile dictobject.c and relink.

The problem is that there *are* different types of changes that require regenerating and recompiling the deepfrozen files, e.g. if the bytecode compiler changes or any of the affected .py modules is changed. But it is difficult to distinguish between the different types of dependencies.

Reply:
> The problem is that there _are_ different types of changes that require regenerating and recompiling the deepfrozen files, e.g. if the bytecode compiler changes or any of the affected .py modules is changed. But it is difficult to distinguish between the different types of dependencies.

Every C code change of a core file triggers a rebuild of the bootstrap interpreter, which then triggers a refreeze of every Python file. Makefile also has order only prerequisites that don't trigger a rebuild. Try this:

```
Python/deepfreeze/importlib._bootstrap.c: $(srcdir)/Lib/importlib/_bootstrap.py $(srcdir)/Tools/scripts/deepfreeze.py $(PYTHON_HEADERS) | $(BOOTSTRAP)
```

The role only triggers a rebuild when a header files changes, but not when the bootstrap interpreter is updated. It still ensure that the bootstrap interpreter is built before the target is executed.

Reply:
This seems to be a [GNU Make feature](https://www.gnu.org/software/make/manual/html_node/Prerequisite-Types.html). Is that a problem for portability? It seems to work on Mac, but IIRC there are some minor (but supported) platforms with less featureful Make implementations.

Also, the definition of `PYTHON_HEADERS` is later in the file than my usage, and then that dependency seems to have no effect on this rule -- I discovered that if I touch `Include/Python.h`, it rebuilds `_bootstrap_python` but doesn't redo the deep-freezing. (But if I touch e.g. `Lib/os.py`, that file does get deep-frozen, so the rest of the rule does work.)

I think I'll keep this idea in mind but I'll first focus on fixing the failing tests, so I can then move on to running the buildbots.